iwyu_output.cc

```diff
@@ -1426,6 +1426,71 @@ void CalculateIwyuForFullUse(OneUse* use,
   }
 }
 
+// This function removes all OneIncludeOrForwardDeclareLine's a that are desired
+// but not present but transitively included by some
+// OneIncludedOrForwardDeclareLine b that is desired and present
+// and adds a's symbol counts to b.
+vector<OneIncludeOrForwardDeclareLine> PruneTransitivelyPresent(
+    vector<OneIncludeOrForwardDeclareLine> const& lines,
+    const IwyuPreprocessorInfo* preprocessor_info) {
+  // In the first step, we copy all lines that
+  // * are either no include lines,
+  // * or that are an include file and present,
+  // * or that are an include file, not present and not desired,
+  // * or that are an include file, not present, desired, and whose include_file
+  //   isn't transitively included by another line which is an include file
+  // into one range and the remaining ones into another range.
+  const auto pred = [&lines, preprocessor_info](
+                        OneIncludeOrForwardDeclareLine const& target) {
+    if (!target.IsIncludeLine() || target.is_present() ||
+        !target.is_desired()) {
+      return true;
+    }
+    // Check if target is transitively included by some source.
+    // This must be proper, equality does not suffice.
+    for (const OneIncludeOrForwardDeclareLine& source : lines) {
+      if (!source.IsIncludeLine()) {
+        continue;
+      }
+      if (source.included_file() == target.included_file()) {
+        continue;
+      }
+
+      if (preprocessor_info->FileTransitivelyIncludes(source.included_file(),
+                                                      target.included_file())) {
+        return false;
+      }
+    }
+    return true;
+  };
+  vector<OneIncludeOrForwardDeclareLine> first, second;
+  first.reserve(lines.size());
+  second.reserve(lines.size());
+  std::partition_copy(lines.cbegin(), lines.cend(), std::back_inserter(first),
+                      std::back_inserter(second), pred);
+  // Now add the symbols from range second to the corresponding elements of
+  // range first.
+  for (const OneIncludeOrForwardDeclareLine& target : second) {
+    const auto iter =
+        std::find_if(first.begin(), first.end(),
+                     [&target, preprocessor_info](
+                         const OneIncludeOrForwardDeclareLine& source) {
+                       return preprocessor_info->FileTransitivelyIncludes(
+                           source.included_file(), target.included_file());
+                     });
+    CHECK_(iter != first.end() &&
+           "Internal error, cannot find the source for a transitive tolerate "
+           "target");
+    for (const auto& p : target.symbol_counts()) {
```

Comment:
Better name than `p` -- `symbol_count`, maybe?

Reply:
I went for `p` since it's a pair because we're iterating over a map.  Other names always have the problem that they somewhat obfuscate we're dealing with a key-value-pair.  But I can change it if you think it makes it more readable.